### ISSUE TITLE
//@formatter:off

### ISSUE THREAD

**Issue by nortakales** [redhat-developer/vscode-java#2397]
This is probably a new feature request rather than a bug.

Currently, `// @formatter:off` and `// @formatter:on` are great for having VSCode not format part of your file. This behavior could also extend to save actions, not just formatting. For example, when using Args4J, you annotate fields with `@Option` and those fields are set for you based on command line arguments (through some kind of reflection black magic). If one of those fields is `final`, then Args4J throws an exception. Example of a field with the annotation:

```java
@Option(name = "--write-tps", handler = LongOptionHandler.class, usage = "The target write tps when updating the table")
private Long writeTps = 1L;
```

When saving, VSCode will add final:

```java
@Option(name = "--write-tps", handler = LongOptionHandler.class, usage = "The target write tps when updating the table")
private final Long writeTps = 1L;
```

but that is not desirable in this case.

Ideally we could use comments to turn off adding final, like this:

```java
// @formatter:off
@Option(name = "--read-capacity", handler = LongOptionHandler.class, usage = "The target read capacity to utilize when scanning the table")
private Long readCapacity = 1L;

@Option(name = "--write-tps", handler = LongOptionHandler.class, usage = "The target write tps when updating the table")
private Long writeTps = 1L;

@Option(name = "--no-dry-run", handler = BooleanOptionHandler.class, usage = "If set to true, the backfill will be done to records needing it")
private Boolean noDryRun = false;
// @formatter:on
```

**Comment by CsCherrYY** [comment]
`@formatter:off` is a part of formatter configurations in eclipse, since vscode-java uses eclipse formatter to formatter your java file, you can change the formatter settings via the formatter profile, ref: https://github.com/redhat-developer/vscode-java/wiki/Formatter-settings

In your formatter profile, you can set `org.eclipse.jdt.core.formatter.enabling_tag` to `@formatter:on` and `org.eclipse.jdt.core.formatter.disabling_tag` to `@formatter:off` to enable your mentioned feature.

**Comment by CsCherrYY** [comment]
I can't reproduce this (auto add `final` when save Java file) even I enabled setting `editor.formatOnSave`.

As far as I know, the extension will add final modifier via source action `Change modifiers to final where possible`, not formatting. You can manually format your file, and check if there are extra finals.

**Comment by nortakales** [comment]
I understand that auto adding the `final` keyword is *not* part of source formatting, but rather is part of `editor.codeActionsOnSave`. Example settings turning that on:

```json
    "[java]": {
        "editor.codeActionsOnSave": {
            "source.generate.finalModifiers": true
        },
    },
```

So, this feature might make more sense with its own special comment rather than using the same as the formatter (i.e. `// @formatter:off`). I thought maybe the same one would make the most sense from a user perspective, but perhaps not. Maybe something like `// @codeActions:off` (for all code actions on save) or `// @final:off` (for just `final`s, though that seems way too specific).

The end result should be that we have *some* comment we can use to avoid automatic code actions on a specific range of lines within a file.

**Comment by rgrunber** [comment]
Makes sense. The code changes on save are known as cleanups. They're basically the same kinds of changes one makes when applying a code action. When we start to support many more, it's likely people will want a way to avoid applying them in a small set of circumstances. Something like a `// @codeActions:{off,on}` would be nice down the road.

I think the place for this feature is probably in JDT Core.

Also, apparently the option itself is [configurable](https://github.com/eclipse-jdt/eclipse.jdt.core/blob/master/org.eclipse.jdt.core/formatter/org/eclipse/jdt/core/formatter/DefaultCodeFormatterConstants.java#L1738-L1742).
